client/asset/dcr/spv_mixing.go

```diff
@@ -0,0 +1,136 @@
+package dcr
+
+import (
+	"context"
+	"fmt"
+
+	"decred.org/dcrwallet/v3/errors"
+	"decred.org/dcrwallet/v3/wallet/udb"
+)
+
+const (
+	smalletCSPPSplitPoint = 1 << 18 // 000.00262144
+	mixedAccountName      = "mixed"
+	mixedAccountBranch    = udb.InternalBranch
+	tradingAccount        = "dextrading"
+)
+
+// checkMixingAccounts checks if the mixed, unmixed and trading accounts
+// required to use this wallet for funds mixing are created and optionally
+// creates any non-existing account. If the accounts exist or are created, the
+// wallet is marked as having mixing accounts and the account numbers for the
+// mixed and unmixed accounts are returned. If any of the accounts do not exist
+// and is not created, an error is returned.
+func (w *spvWallet) checkMixingAccounts(ctx context.Context, create bool) (uint32, uint32, error) {
+	// returns an error if the account doesn't exist and is not created.
+	checkAccount := func(acct string) (uint32, error) {
+		acctNum, err := w.dcrWallet.AccountNumber(ctx, acct)
+		if errors.Is(err, errors.NotExist) && create {
+			acctNum, err = w.dcrWallet.NextAccount(ctx, acct)
+		}
+		return acctNum, err
+	}
+
+	unmixedAcctNum := defaultAcct // the default acct is the unmixed account
+
+	mixedAcctNum, err := checkAccount(mixedAccountName)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	_, err = checkAccount(tradingAccount) // don't need the account number, but should exist as well
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// If we got here, all required accounts exist.
+	w.hasMixedAccount.Store(true)
+	return mixedAcctNum, uint32(unmixedAcctNum), nil
+}
+
+// IsMixing is true if the wallet is currently mixing funds.
+func (w *spvWallet) IsMixing() bool {
+	w.csppMtx.Lock()
+	defer w.csppMtx.Unlock()
+	return w.cancelCSPPMixer != nil
+}
+
+// StartFundsMixer starts mixing funds. Creates the required accounts (including
+// the dex trading account) if any does not already exist.
+func (w *spvWallet) StartFundsMixer(ctx context.Context, passphrase []byte) error {
```

Comment:
This is currently called when the wallet is started or the config is updated, if the wallet config requires starting the mixer automatically. However, in both those cases, the wallet passphrase isn't available and this would error if the wallet wasn't already unlocked. On app start, the wallet is always locked, so it appears we can't support automatic mixing on start. Leaving the support for now to see if we could work around it - mixing as soon as the wallet is synced, is a good thing.

Reply:
There could be an "auto unlock" setting for this wallet, which if set to true, automatically unlocks the wallet on login. Maybe it is only left unlocked if there are unmixed outputs that need to be mixed. Deposit page for wallets with mixing enabled could also have a message letting users know they need to unlock the wallet, if it isn't already, so that their deposit gets mixed.

Reply:
> There could be an "auto unlock" setting for this wallet, which if set to true, automatically unlocks the wallet on login. Maybe it is only left unlocked if there are unmixed outputs that need to be mixed. Deposit page for wallets with mixing enabled could also have a message letting users know they need to unlock the wallet, if it isn't already, so that their deposit gets mixed.

Maybe something like https://github.com/buck54321/dcrdex/compare/916fc36...buck54321:dcrdex:login-start-funds-mixers?

Reply:
Yeah, I think that's good.

Reply:
I'm just going to cherry-pick that. Looks neat.